### PR TITLE
add func of plotting losses

### PR DESCRIPTION
**Pull request by lijialin03**
### PR types

New features
### PR changes

Others
### Describe

1. add a function to plot train losses
2. usage like,
image

epoch loss of deeponet's constraint with smooth_step=1, use_semilogy=False:
image
with smooth_step=1000, use_semilogy=False:
image
with smooth_step=1, use_semilogy=True:
image
with smooth_step=1000, use_semilogy=True:
image

### CODE REVIEW

File: ppsci/solver/solver.py [hunk 1]
```diff
@@ -298,6 +298,11 @@ def __init__(
         jit.enable_to_static(to_static)
         logger.info(f"Set to_static={to_static} for forward computation.")
 
+        # recode losses
+        self.losses_dict = {"loss": []}
+        for key in self.constraint.keys():
+            self.losses_dict[key] = []
+
```
Comment (HydrogenSulfate) [inline ppsci/solver/solver.py:305]: 这个记录器loss的变量可以做到`misc.AverageMeter`这个类里面，比如给它加一个属性 `history: Dict[str, List[float]]`

File: ppsci/solver/solver.py [hunk 2]
```diff
@@ -671,3 +676,18 @@ def no_sync_context_manager(
                 else contextlib.suppress()
             )
         return ctx_manager
+
+    def plot_losses(self, by_epoch: bool = False, smooth_step: int = 1) -> None:
```
Comment (HydrogenSulfate) [inline ppsci/solver/solver.py:680]: plot_loss_history

File: ppsci/solver/train.py [hunk 3]
```diff
@@ -98,6 +98,10 @@ def train_epoch_func(solver: "solver.Solver", epoch_id: int, log_freq: int):
             else:
                 total_loss.backward()
 
+            # record losses
+            for key in loss_dict:
+                solver.losses_dict[key].append(loss_dict[key])
+
```
Comment (HydrogenSulfate) [inline ppsci/solver/train.py:104]: 可以放到`printer.update_train_loss`函数里去

File: ppsci/utils/misc.py [hunk 4]
```diff
@@ -299,3 +302,55 @@ def function_with_eval_state(self, *args, **kwargs):
...
+) -> None:
+    """Plotting loss-iteration/epoch curve.
+
+    Args:
+        loss_dict (Dict[str, List]): Losses of all constraints.
+        output_dir (str): Output directory. Defaults to "./output/".
+        by_epoch (bool, optional): Whether the abscissa axis of the curve is epoch or iteration. Defaults to False.
+        iters_per_epoch (int, optional): Number of iterations within an epoch. Defaults to 1.
+        smooth_step (int, optional): How many steps of loss are squeezed to one point to smooth the curve. Defaults to 1.
+    """
```
Comment (HydrogenSulfate) [inline ppsci/utils/misc.py:322]: 在misc下面的函数考虑做得更通用一点，比如做成`plot_curve`，而不是专门为了画loss，这样的话后续想画 eval metric、eval loss都可以调用这个函数

File: ppsci/utils/misc.py [hunk 5]
```diff
@@ -299,3 +302,55 @@ def function_with_eval_state(self, *args, **kwargs):
...
 
     return function_with_eval_state
+
+
+def plot_losses_fig(
+    loss_dict: Dict[str, List],
+    output_dir: str = "./output/",
+    by_epoch: bool = False,
+    iters_per_epoch: int = 1,
+    smooth_step: int = 1,
```
Comment (HydrogenSulfate) [inline ppsci/utils/misc.py:312]: 接口设计可以参考下方代码：
``` py
def plot_curve(
    loss_history: Dict[str, List[float]],
    fname: str,
    by_epoch: bool = False,
    iters_per_epoch: int = 1,
    smooth_step: int = 1,
):
    ...
```

**Review by HydrogenSulfate — CHANGES_REQUESTED** [review]

**Review by HydrogenSulfate — CHANGES_REQUESTED** [review]
1. loss记录功能跟printer模块以及AverageMeter类进行合并
2. 重新进行合理的功能拆分，减少耦合，misc下的函数尽可能做得通用，前面大部分的预处理代码移动到Solver.plot_loss_history里面

File: ppsci/utils/misc.py [hunk 6]
```diff
@@ -299,3 +304,49 @@ def function_with_eval_state(self, *args, **kwargs):
...
+
+    # smooth
+    if data_arr.shape[0] % smooth_step != 0:
+        data_vis = np.reshape(
+            data_arr[: -(data_arr.shape[0] % smooth_step), :],
+            (-1, smooth_step, data_arr.shape[1]),
+        )
+    else:
+        data_vis = np.reshape(data_arr, (-1, smooth_step, data_arr.shape[1]))
+    data_vis = np.mean(data_vis, axis=1)
```
Comment (HydrogenSulfate) [inline ppsci/utils/misc.py:336]: 直接使用data_arr就行了，不需要一个新的data_vis变量

File: ppsci/utils/misc.py [hunk 7]
```diff
@@ -299,3 +304,49 @@ def function_with_eval_state(self, *args, **kwargs):
...
+
+    # plot
+    plt.figure()
+    for i in range(data_vis.shape[1]):
+        plt.semilogy(np.arange(data_vis.shape[0]) * smooth_step, data_vis[:, i])
```
Comment (HydrogenSulfate) [inline ppsci/utils/misc.py:341]: 这个是必要的吗，取对数作为默认画图配置感觉不是很合理

File: ppsci/utils/misc.py [hunk 8]
```diff
@@ -299,3 +304,49 @@ def function_with_eval_state(self, *args, **kwargs):
...
+    plt.legend(
+        list(data.keys()),
+        loc="lower left",
+    )
+    plt.xlabel(fname[0])
+    plt.ylabel(fname[1])
```
Comment (HydrogenSulfate) [inline ppsci/utils/misc.py:347]: 不使用fname

**Review by HydrogenSulfate — CHANGES_REQUESTED** [review]

**Comment by HydrogenSulfate** [comment]
可以使用DeepONet之类的训练速度比较快的案例作为展示，把保存的loss曲线图放到PR描述里

File: ppsci/solver/solver.py [hunk 9]
```diff
@@ -673,26 +673,31 @@ def no_sync_context_manager(
             )
         return ctx_manager
 
-    def plot_loss_history(self, by_epoch: bool = False, smooth_step: int = 1) -> None:
+    def plot_loss_history(
+        self, by_epoch: bool = False, smooth_step: int = 1, use_semilogy=False
+    ) -> None:
```
Comment (HydrogenSulfate) [inline ppsci/solver/solver.py:678]: 要不这里的use_semilogy默认为True吧，这个函数比较特殊，打印train loss

**Review by HydrogenSulfate — APPROVED** [review]
LGTM
